File: web_sdk/src/transport/connection-manager.ts
// src/transport/connection-manager.ts

import { TypedEventEmitter } from '../utils/typed-emitter.js';
import { WsManager } from './ws-manager.js';
import { sleep } from '../utils/retry.js';
import type { Logger } from '../utils/logger.js';

export type ConnectionState = 'connecting' | 'connected' | 'reconnecting' | 'disconnected';

const BACKOFF_DELAYS = [1000, 2000, 4000, 8000, 16000, 32000, 60000];

/**
 * Manages WebSocket connection lifecycle with exponential backoff reconnect.
 */
export class ConnectionManager {
  state: ConnectionState = 'disconnected';
  readonly stateChanges = new TypedEventEmitter<{ change: { state: ConnectionState } }>();

  private reconnecting = false;
  private destroyed = false;
  private connectResolvers: Array<() => void> = [];
  private backoffIndex = 0;

  constructor(
    private readonly wsManager: WsManager,
    private readonly getWsUrl: () => string,
    private readonly logger: Logger,
  ) {
    wsManager.events.on('open', () => {
      this.backoffIndex = 0;
      this.setState('connected');
      // Resolve all pending connect() promises
      const resolvers = this.connectResolvers.splice(0);
      for (const r of resolvers) r();
    });

    wsManager.events.on('close', ({ code }) => {
      if (this.destroyed) return;
      // Normal close (1000) only if we initiated it
      if (this.state !== 'disconnected') {
        this.scheduleReconnect();
      }
    });

    wsManager.events.on('error', () => {
      if (this.destroyed) return;
      if (this.state !== 'disconnected') {
        this.scheduleReconnect();
      }
    });
  }

  async ensureConnected(): Promise<void> {
    if (this.state === 'connected') return;
    if (this.state === 'connecting' || this.reconnecting) {
      return new Promise<void>((resolve) => {
        this.connectResolvers.push(resolve);
      });
    }
    this.doConnect();
    return new Promise<void>((resolve) => {
      this.connectResolvers.push(resolve);
    });
  }

  async disconnect(): Promise<void> {
    this.reconnecting = false;
    this.setState('disconnected');
    this.wsManager.disconnect();
  }

  destroy(): void {
    this.destroyed = true;
    this.wsManager.disconnect();
    this.setState('disconnected');
  }

  private doConnect(): void {
    this.setState('connecting');
    const url = this.getWsUrl();
    this.wsManager.connect(url);
  }

  private scheduleReconnect(): void {
    if (this.reconnecting || this.destroyed) return;
    this.reconnecting = true;
    this.setState('reconnecting');

    const delay = BACKOFF_DELAYS[Math.min(this.backoffIndex, BACKOFF_DELAYS.length - 1)] ?? 60000;
    this.backoffIndex++;

    this.logger.info(`Reconnecting in ${delay}ms`);

    sleep(delay).then(() => {
      if (this.destroyed) return;
      this.reconnecting = false;
      this.doConnect();
    });
  }

  private setState(state: ConnectionState): void {
    this.state = state;
    this.stateChanges.emit('change', { state });
  }
}

File: flutter_sdk/lib/src/transport/connection_manager.dart
import 'dart:async';
import '../utils/logger.dart';
import 'ws_channel.dart';

/// WebSocket connection state.
enum ConnectionState {
  connecting,
  connected,
  reconnecting,
  disconnected,
}

/// Manages the WebSocket connection lifecycle with exponential backoff.
///
/// Backoff schedule: 1s → 2s → 4s → 8s → 16s → 32s → 60s (cap).
class ConnectionManager {
  final WsChannel _wsChannel;
  final String Function() _wsUrlBuilder;
  final SdkLogger _logger;

  final StreamController<ConnectionState> _stateController =
      StreamController<ConnectionState>.broadcast();

  ConnectionState _state = ConnectionState.disconnected;
  bool _disposed = false;
  int _reconnectAttempts = 0;
  Timer? _reconnectTimer;
  StreamSubscription? _wsErrorSubscription;

  static const _backoffDelays = [1, 2, 4, 8, 16, 32, 60];

  ConnectionManager({
    required WsChannel wsChannel,
    required String Function() wsUrlBuilder,
    SdkLogger? logger,
  })  : _wsChannel = wsChannel,
        _wsUrlBuilder = wsUrlBuilder,
        _logger = logger ?? SdkLogger(tag: 'ConnectionManager');

  /// Stream of connection state changes.
  Stream<ConnectionState> get state => _stateController.stream;

  /// Current connection state.
  ConnectionState get currentState => _state;

  /// Ensures the WebSocket is connected. If already connected, no-op.
  Future<void> ensureConnected() async {
    if (_state == ConnectionState.connected) return;
    await _connect();
  }

  /// Manually disconnect. Disables auto-reconnect until [ensureConnected] is called again.
  Future<void> disconnect() async {
    _setState(ConnectionState.disconnected);
    _reconnectTimer?.cancel();
    _wsErrorSubscription?.cancel();
    await _wsChannel.disconnect();
  }

  /// Clean up all resources.
  Future<void> dispose() async {
    _disposed = true;
    await disconnect();
    await _stateController.close();
  }

  // ---------------------------------------------------------------------------
  // Private
  // ---------------------------------------------------------------------------

  Future<void> _connect() async {
    _setState(ConnectionState.connecting);
    _wsErrorSubscription?.cancel();

    try {
      final wsUrl = _wsUrlBuilder();
      await _wsChannel.connect(wsUrl);
      _setState(ConnectionState.connected);
      _reconnectAttempts = 0;

      // Listen for disconnection via stream completing/erroring
      _wsErrorSubscription = _wsChannel.events.listen(
        null,
        onError: (_) => _onDisconnect(),
        onDone: _onDisconnect,
      );
    } catch (e) {
      _logger.error('Connection failed: $e');
      _scheduleReconnect();
    }
  }

  void _onDisconnect() {
    if (_disposed || _state == ConnectionState.disconnected) return;
    _logger.info('WebSocket disconnected — scheduling reconnect');
    _scheduleReconnect();
  }

  void _scheduleReconnect() {
    if (_disposed) return;
    _setState(ConnectionState.reconnecting);
    _reconnectTimer?.cancel();

    final delaySeconds = _backoffDelays[
        _reconnectAttempts.clamp(0, _backoffDelays.length - 1)];
    _reconnectAttempts++;

    _logger.info('Reconnecting in ${delaySeconds}s (attempt $_reconnectAttempts)');

    _reconnectTimer = Timer(Duration(seconds: delaySeconds), () {
      if (!_disposed) {
        _connect();
      }
    });
  }

  void _setState(ConnectionState newState) {
    if (_state == newState) return;
    _state = newState;
    _logger.debug('Connection state: $newState');
    if (!_stateController.isClosed) {
      _stateController.add(newState);
    }
  }
}
